pkg/declcfg/bundle.go

```diff
@@ -0,0 +1,30 @@
+package declcfg
+
+import (
+	"github.com/operator-framework/operator-registry/pkg/property"
+)
+
+type bundle struct {
```

Comment:
I'm gonna be honest here by saying that I don't see the value of this Bundle API given it is pretty much identical to Bundle on registry package. The diff I can see here is the registry.Bundle has more fields and this bundle has `RelatedImages`. I understand this is helpful for the validation but I wonder if we reinvent the wheel here because I can see we potentially add more fields and it will become exactly like registry.Bundle.

Reply:
I recall our conversation about moving this to api repo. I think that makes more sense.

Reply:
I disagree about re-using `registry.Bundle` in place of the declarative config `bundle` struct. `declcfg.bundle` is the Go specification of the `olm.bundle` schema. It is used to (un)marshal declarative config JSON. That should be its singular purpose.

Looking at `registry.Bundle`, I get the impression that its primary purpose is the representation of the _contents_ of the bundle image, used when we unpack the bundle image and need to understand what it contains.

IMO using one struct to represent both declarative config's `olm.bundle` schema AND the contents of bundle described by that schema will cause unnecessary complexity. For instance, if we need to add a new field to `registry.Bundle` to capture some detail about the contents of the bundle, we also then need to understand how that might impact that struct's use as the specification for the `olm.bundle` spec.

Reply:
I theory, I would agree that `operator-framework/api` is a logical place for `pkg/declcfg`. I think my suggestion for now would be to keep it all encapsulated in `operator-registry` so that we can iterate on it (if necessary) without involving another repository.

We can always move it `operator-framework/api` once it stabilizes. To prevent others from making use of it for now, we could put it in an `internal` tree to make it available only to other `operator-registry` packages.

Related side question: does `operator-registry` make any Go API guarantees? Given that there are v1.x tags, I would assume that we're not allowed to remove or break any exported APIs?

Reply:
> I disagree about re-using `registry.Bundle` in place of the declarative config `bundle` struct. `declcfg.bundle` is the Go specification of the `olm.bundle` schema. It is used to (un)marshal declarative config JSON. That should be its singular purpose.
> 
> Looking at `registry.Bundle`, I get the impression that its primary purpose is the representation of the _contents_ of the bundle image, used when we unpack the bundle image and need to understand what it contains.
> 
> IMO using one struct to represent both declarative config's `olm.bundle` schema AND the contents of bundle described by that schema will cause unnecessary complexity. For instance, if we need to add a new field to `registry.Bundle` to capture some detail about the contents of the bundle, we also then need to understand how that might impact that struct's use as the specification for the `olm.bundle` spec.

1. I understand your intention to use this `declcfg.bundle` as the representation of `olm.bundle` JSON struct in the config and you are correct regarding `registry.bundle` is the representation of the bundle and its content. However, you blur the line between the two when you include `CsvJSON` and `Objects` in the `declcfg.bundle`. Those two fields are for bundle content so you are duplicating the `registry.bundle` purpose. I can see the comment you have on those two fields about preserving the compatibility with gRPC which actually colliding with another API we have also named Bundle (`api.Bundle` which is for gRPC API). As you can see, we end up with 3 different `something.Bundle` that represent essentially the same thing. I'm not saying this `declcfg.bundle` is a bad thing given we can't really change `api.Bundle` because that would break bundle gRPC API and `registry.Bundle` is kind of a mess due to legacy reason (we don't want to remove fields and keep adding more as we introduced `property` and `dependency` for resolution). So I can see this API perhaps is a necessary step to move forward with config work without dealing with existing API. However, it seems to become a troubling pattern that we have 3 APIs with the same name floating around in the same repo that represents the same thing perhaps in slightly different shapes or forms. Imagining someone looking our registry codebase and wonder why there are 3 different bundles here and frankly I wonder if they even notice these are 3 different APIs or they just assume it is the same thing.
2. We can decide where to keep this bundle API at a later time. I don't feel strongly about where it should belong though as I mentioned above, perhaps it should be in api for the sake of not having multiple bundle APIs in the same repo. Plus, it depends on who we think who is dictating the config format. I often think personally api repo is where OLM and registry have in common and we have everything there so they can share and reuse. In term of config, OLM doesn't use it directly so the config API is actually being dictated by registry which means it can stay here. There is nothing wrong with folks using API from registry repo directly.
3. I don't know if we guarantee the backward compatibility for any API registry except the gRPC API for bundle. The rest is pretty much for internal use. No one is using the APIs in registry directly as far as I know.

Reply:
1. Pretty much agree with everything you said. Perhaps we could beef up the GoDoc comments for each `Bundle` type to try to clarify where/why they're used?
2. I think we need to consider `operator-sdk` in the list of clients of `operator-framework/api`. I haven't thought about it _too_ much, but I could imagine `operator-sdk` needing to manipulate declarative configs at some point. Either way, my vote would be to leave it here for now as an API that can only be used internally by `operator-registry` components.

    Based on this, I'm thinking I'll move `declcfg` and `model` into an `internal` package so that other projects will be prevented from importing it. Then if/when there's a need for others to consume it, we can have this conversation again. And by then we'll have a concrete motivation to help us decide. WDYT?
3. This point is going off on a tangent, but I know [SDK imports `operator-registry`](https://github.com/operator-framework/operator-sdk/blob/b3547bcacaca9e10c7225ced6ebd6210067d7fc6/go.mod#L16), so there's at least one consumer of "internal" APIs. We should move everything that's actually internal to `internal` directory trees to prevent other projects from using them.